core/src/data_sources/data_source.rs

```diff
@@ -594,23 +596,99 @@ impl DataSource {
+                    limit: Some(1000),
+                    offset: page_offset,
+                    filter: Some(qdrant::Filter {
+                        must_not: vec![],
+                        should: vec![],
+                        must: vec![qdrant::FieldCondition {
+                            key: "document_id_hash".to_string(),
+                            r#match: Some(qdrant::Match {
+                                match_value: Some(qdrant::r#match::MatchValue::Text(
+                                    document_id_hash.clone(),
+                                )),
+                            }),
+                            ..Default::default()
+                        }
+                        .into()],
+                    }),
+                    ..Default::default()
+                })
+                .await?;
+
+            scroll_results
+                .result
+                .iter()
+                .for_each(|result| match result.payload.get("text") {
+                    Some(t) => match t.kind {
+                        Some(qdrant::value::Kind::StringValue(ref chunk_text)) => {
+                            match result.vectors.clone() {
+                                Some(v) => match v.vectors_options {
+                                    Some(VectorsOptions::Vector(v)) => {
+                                        let mut hasher = blake3::Hasher::new();
+                                        hasher.update(chunk_text.as_bytes());
+                                        let text_only_hash =
+                                            format!("{}", hasher.finalize().to_hex());
+
+                                        texthash2embedding.insert(
+                                            text_only_hash.to_string(),
+                                            EmbedderVector {
+                                                created: document.created,
+                                                vector: v
+                                                    .data
+                                                    .iter()
+                                                    .map(|v| *v as f64)
+                                                    .collect::<Vec<f64>>(),
+                                                model: self.config.model_id.clone(),
+                                                provider: self.config.provider_id.to_string(),
+                                            },
+                                        );
+                                    }
+                                    _ => {}
+                                },
+                                None => {}
+                            }
+                        }
+                        _ => {}
+                    },
+                    None => {}
+                });
+            page_offset = scroll_results.next_page_offset;
+            if page_offset.is_none() {
```

Comment:
I'm a bit antsy when I see a loop { }

given that this is caching and we take 1024 chunks at once, can we add a counter to exist if we loop more than 8 times?

Reply:
I'm a bit more worried about an infinite loop in core vs connectors.

To be clear, this is defense against qdrant not being correct in its response.

Feel like it's a minimal diff that will give us peace of mind at low cost?